projects/dendrites/gaussian_classification/run_dendritic_network.py

```diff
@@ -49,26 +49,64 @@
 from projects.dendrites.gaussian_classification.gaussian import GaussianDataset
 
 
-def hardcode_dendritic_weights(dendrite_segments, context_vectors):
-    """
-    We hardcode the weights of dendrites such that each unit recognizes a single random
-    context vector. The first dendritic segment is initialized to contain positive
-    weights from that context vector. The other segment(s) ensure that the unit is
-    turned off for any other context - they contain negative weights for all other
-    weights.
-    """
+def hardcode_dendritic_weights(dendrite_segments, context_vectors, init):
     num_units, num_segments, dim_context = dendrite_segments.weights.size()
     num_contexts, _ = context_vectors.size()
-    fixed_segment_weights = torch.zeros((num_units, num_segments, dim_context))
 
-    for i in range(num_units):
-        context_perm = context_vectors[torch.randperm(num_contexts), :]
-        fixed_segment_weights[i, :, :] = 1.0 * (context_perm[0, :] > 0)
-        fixed_segment_weights[i, 1:, :] = -1
-        fixed_segment_weights[i, 1:, :] += fixed_segment_weights[i, 0, :]
-        del context_perm
+    if init == "overlapping":
+        # We hardcode the weights of dendrites such that each context selects 5% of
+        # hidden units to become active and form a subnetwork. Hidden units are sampled
+        # with replacement, hence subnetworks can overlap. Any context/task which does
+        # not use a particular hidden unit will cause it to turn off, as the unit's
+        # other segment(s) have -1 in all entries and will yield an extremely small
+        # dendritic activation.
+
+        new_dendritic_weights = -0.95 * torch.ones((num_units, num_segments,
+                                                    dim_context))
+
+        # The number of units to allocate to each context (with replacement)
+        k = int(0.05 * num_units)
+
+        # Keep track of the number of contexts for which each segment has already been
+        # chosen; this is to not overwrite a previously hardcoded segment
+        num_contexts_chosen = {i: 0 for i in range(num_units)}
+
+        for c in range(num_contexts):
+
+            # Pick k random units to be activated by the cth context
+            selected_units = torch.randperm(num_units)[:k]
+            for i in selected_units:
+                i = i.item()
+
+                # If num_segments other contexts have already selected unit i to become
+                # active, skip
+                segment_id = num_contexts_chosen[i]
```

Comment:
Because of the 'if' on line 84, this loop could allocate fewer than k units for a given context, right?

Reply:
That's correct, but that would only occur if there are fewer segments per unit than number of tasks. In the current config, the number of segments is set to the number of tasks, so the 'if' on line 84 won't be triggered.

Reply:
Ok, makes sense.